devicetypes/smartthings/fidure-2nd-generation-thermostat.src/fidure-2nd-generation-thermostat.groovy

```diff
@@ -0,0 +1,810 @@
+		attrId = 0x11
+		currentSetPoint = device.currentValue("coolingSetpoint")
+		attrName = "coolingSetpoint"
+	} else return
+
+	// move up and down in exact steps of C or F
+	// clean up the current set point
+	if (getTemperatureScale() == "C") {
+		value = (value / 10)
+		currentSetPoint = roundtoHalves(currentSetPoint)
+	}else { // Fahrenheit
+		currentSetPoint = Math.round(currentSetPoint)
+		value = (value / 5)
+	}
+	
+	currentSetPoint += value
+
+	if (!checkBoundary(mode, currentSetPoint))
+		return;
+	
+    device.updateDataValue(attrName, "$currentSetPoint")
+	sendEvent(name: attrName, value: currentSetPoint)
+    
+	def cmds = setSetpoint(attrId, currentSetPoint)
+
+    runIn(3, "set"+ attrName + "AfterDelay" , [overwrite: true, data: [c: cmds]])
+    return null
+}
+
+def checkBoundary(mode, newValue){
+	newValue = getCelsiusValueX100(newValue)
+	def min = Double.parseDouble(device.getDataValue("min" + mode + "Setpoint")?: "800") 
+	def max = Double.parseDouble(device.getDataValue("max" + mode + "Setpoint")?: "3000")
+	
+	if (newValue <= max && newValue >= min)
+		return true;
+	return false;
+}
+
+def sendZigBeCommands(data) {
+	//log.trace "Sending ZigBee Commands: $data?.c[0]"
+	sendZigbeeCmds(data.c[0], 100)
+}
+
+def sendZigbeeCmds(cmds, delay = 2000) {
+    // remove zigbee library added "delay 2000" after each command
+    // the new sendHubCommand won't honor these, instead it'll take the delay as argument
+
+    cmds?.removeAll { it.startsWith("delay") }
+    // convert each command into a HubAction
+    cmds = cmds.collect { new physicalgraph.device.HubAction(it) }
+    //log.trace "hub command: $cmds delay $delay"
+    sendHubCommand(cmds, delay)
+}
+// ST platform currently does not have a time service.  This command allows for a smartApp to Manually update the time.
+def setThermostatTime() {
+  	Date date = new Date()
+
+	long millis = date.getTime() // Millis since Unix epoch
+  	millis -= 946684800000  // adjust for ZigBee EPOCH
+	// adjust for time zone and DST offset
+	millis += location.timeZone.getOffset(date.getTime())
+	//convert to seconds
+	millis /= 1000
+	
+	// POSSIBLE BUG:
+	// write attribute does not pack int32u properly.  so we need to revese the endianness
+	millis = zigbee.convertHexToInt(DataType.pack(millis, DataType.UINT32, true))
+		
+	zigbee.writeAttribute(0x201, 0x000F, 0x23, millis, [mfgCode: 0x1121])
+}
+
+def updateTime() {
+    sendZigbeeCmds(setThermostatTime(), 100)
+}
+
+def poll() {
+	log.trace "Poll..."
+    refresh()
+}
+
+def lock(){	setLockLevel(getLockLevelMap()[settings.lock_level] ?: 5) }
+def unlock() { setLockLevel(0) }
+
+def setLockLevel(Integer level) {
+	if (level < 0 || level > 5 )
+		return
+   
+	[ 
+	zigbee.writeAttribute(0x0204, 0x0001 , 0x30 , level) ,
+	"delay 300",
+	zigbee.readAttribute(0x0204, 0x0001)
+	] 
+}
+
+// ----------------
+// helper functions 
+// ----------------
+def getTemperature(value) { 
+	Double celsius = Integer.parseInt(value, 16) / 100
```

Comment:
consider to validate the value before parsing it, unless you for sure now the thermostat won't send anything unexpected. Please compare to the Zen Thermostat implementation